llvm/lib/Support/raw_ostream.cpp

```diff
@@ -563,6 +563,69 @@ void format_object_base::home() {
 //===----------------------------------------------------------------------===//
 //  raw_fd_ostream
 //===----------------------------------------------------------------------===//
+static void emitExtraOpenFileFailDiagnostic(const Twine &Name,
+                                            const Twine &Error,
+                                            int Retries) {
+  SmallVector<char, 64> Buffer;
+  raw_svector_ostream OS(Buffer);
+  OS << "Open File failed on file: " << Name << "\n";
+  OS << "Error: " << Error << "\n";
+  OS << "Attempted again: " << Retries << " times\n";
+  StringRef MessageStr = OS.str();
+  ::write(2, MessageStr.data(), MessageStr.size());
+}
+
+static std::error_code openFileForReadWriteWithRetry(const Twine &Name, int &ResultFD,
+                                                     sys::fs::CreationDisposition Disp,
+                                                     llvm::sys::fs::OpenFlags Flags) {
+  // This is a workaround for a Swift compiler issue wherein the compiler
+  // occassionally fails when destructing a file descriptor with no meaningful
+  // diagnostic.
+  // Re-try opening descriptor, in case it helps.
+  // Emit additional diagnostics to ease with root-causing this issue.
+  // Reversal tracked in: rdar://74359658
+  auto EC = sys::fs::openFileForReadWrite(Name, ResultFD, Disp, Flags);
+
+  if (EC) {
+    const unsigned MAX_COUNT = 10;
+    unsigned I = 0;
+    for (I = 0; I != MAX_COUNT; ++I) {
+      auto RetryEC = sys::fs::openFileForReadWrite(Name, ResultFD, Disp, Flags);
+      if (!RetryEC) {
+        EC = RetryEC;
+        break;
+      }
+    }
+    emitExtraOpenFileFailDiagnostic(Name, EC.message(), I);
+  }
+
+  return EC;
+}
+static std::error_code openFileForWriteWithRetry(const Twine &Name, int &ResultFD,
```

Comment:
could we merge this with `openFileForReadWriteWithRetry` by adding a boolean parameter `WithRead`?

Reply:
Done.